test/jdk/sun/security/pkcs11/KeyStore/ImportKeyToP12.java

```diff
@@ -0,0 +1,136 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.MessageDigest;
+import java.security.Provider;
+import java.security.Security;
+
+import javax.crypto.Cipher;
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+/*
+ * @test
+ * @bug 8301553
+ * @summary test SunPKCS11's password based privacy and integrity
+ *          applied to PKCS #12 keystores
+ * @library /test/lib ..
+ * @modules java.base/sun.security.util
+ * @run main/othervm/timeout=30 ImportKeyToP12
+ */
+
+public final class ImportKeyToP12 extends PKCS11Test {
+    private static final String alias = "alias";
+    private static final char[] password = "123456".toCharArray();
+    private static final Key key = new SecretKeySpec(new byte[] {
+            0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7,
+            0x8, 0x9, 0xa, 0xb, 0xc, 0xd, 0xe, 0xf }, "AES");
+    private static final String[] pbeCipherAlgs = new String[] {
+            "PBEWithHmacSHA1AndAES_128", "PBEWithHmacSHA224AndAES_128",
+            "PBEWithHmacSHA256AndAES_128", "PBEWithHmacSHA384AndAES_128",
+            "PBEWithHmacSHA512AndAES_128", "PBEWithHmacSHA1AndAES_256",
+            "PBEWithHmacSHA224AndAES_256", "PBEWithHmacSHA256AndAES_256",
+            "PBEWithHmacSHA384AndAES_256", "PBEWithHmacSHA512AndAES_256"
+    };
+    private static final String[] pbeMacAlgs = new String[] {
+            "HmacPBESHA1", "HmacPBESHA224", "HmacPBESHA256",
+            "HmacPBESHA384", "HmacPBESHA512"
+    };
+    private static final KeyStore p12;
+    private static final String sep = "======================================" +
+            "===================================";
+
+    static {
+        KeyStore tP12 = null;
+        try {
+            tP12 = KeyStore.getInstance("PKCS12");
+        } catch (KeyStoreException e) {}
+        p12 = tP12;
+    }
+
+    public void main(Provider sunPKCS11) throws Exception {
+        System.out.println("SunPKCS11: " + sunPKCS11.getName());
+        // Test all privacy PBE algorithms with an integrity algorithm fixed
+        for (String pbeCipherAlg : pbeCipherAlgs) {
+            // Make sure that SunPKCS11 implements the Cipher algorithm
+            Cipher.getInstance(pbeCipherAlg, sunPKCS11);
+            testWith(sunPKCS11, pbeCipherAlg, pbeMacAlgs[0]);
```

Comment:
Looks like an error handling is missing, i.e. catch and ignore/skip the corresponding testing if the algorithm is not supported by the sunPKCS11 provider under test. Same goes for the Mac loop below also.

Reply:
I'm not sure of this change because the test is NSS-based —in fact, we initialize the NSS software token with a specific configuration— and the PBE algorithms used in the test should be available. Have you found a platform/NSS version in which this is not true? I'm concerned about making the test (silently) pass when there is something that we should notice, such as the removal of something in a newer version that would affect our testing coverage or the change in one of the native mechanisms for a given algorithm.